### PR TITLE
Enable Mono ARM & ARM64 Linux builds

### PR DESCRIPTION
**Pull request by directhex**
Traditionally, Mono is build with native compilers (i.e. "build for this architecture" running in an ARM or ARM64 chroot on an ARM64 server), but to align with CoreCLR, let's go the cross-compile way.

This enables cross-compiler support, by passing a whole bunch of extra info to `./configure`, some of which is set in the common YAML (like the `--sysroot` value)

### CODE REVIEW

**Review by filipnavara — APPROVED** [review]

**Review by marek-safar — APPROVED** [review]

**Comment by directhex** [comment]
I realise I'm not certain this is building an ARM{64} corlib, I was so focused on getting the runtime to build. Will verify.

**Comment by directhex** [comment]
`                             /home/directhex/Projects/runtime/artifacts/obj/mono/System.Private.CoreLib/arm64/Debug/System.Private.CoreLib.dll`

Seems legit?

File: src/mono/mono.proj [hunk 1]
```diff
@@ -32,9 +32,29 @@
...
     </PropertyGroup>
+    <PropertyGroup Condition=" '$(TargetArchitecture)' == 'arm64' ">
+        <_MonoTuple>aarch64-linux-gnu</_MonoTuple>
+    </PropertyGroup>
+    <PropertyGroup Condition=" '$(TargetArchitecture)' == 'arm' ">
+        <_MonoTuple>arm-linux-gnueabihf</_MonoTuple>
+    </PropertyGroup>
+    <PropertyGroup Condition=" '$(TargetArchitecture)' == 'arm' Or '$(TargetArchitecture)' == 'arm64' ">
+        <_MonoCrossDir>$(ROOTFS_DIR)</_MonoCrossDir>
+        <_MonoExtraCFLAGS>$(_MonoExtraCFLAGS) --sysroot=$(_MonoCrossDir) --target=$(_MonoTuple)</_MonoExtraCFLAGS>
```
Comment (lewurm) [inline src/mono/mono.proj]: what is the purpose of `--target=$(_MonoTuple)` in `CFLAGS` and friends?

File: src/mono/mono.proj [hunk 2]
```diff
@@ -32,9 +32,29 @@
...
+        <_MonoExtraCXXFLAGS>$(_MonoExtraCXXFLAGS) --sysroot=$(_MonoCrossDir) --target=$(_MonoTuple)</_MonoExtraCXXFLAGS>
+        <_MonoExtraLDFLAGS>$(_MonoExtraLDFLAGS) --sysroot=$(_MonoCrossDir) --target=$(_MonoTuple)</_MonoExtraLDFLAGS>
+        <_MonoExtraCCLDFLAGS>$(_MonoExtraCCLDFLAGS) -XCClinker --target=$(_MonoTuple)</_MonoExtraCCLDFLAGS>
+        <_MonoConfigureParams>$(_MonoConfigureParams) --host=$(_MonoTuple) --target=$(_MonoTuple)</_MonoConfigureParams>
+        <_MonoConfigureParams>$(_MonoConfigureParams) RANLIB=$(_MonoTuple)-ranlib</_MonoConfigureParams>
+        <_MonoConfigureParams>$(_MonoConfigureParams) LD=$(_MonoTuple)-ld</_MonoConfigureParams>
+        <_MonoConfigureParams>$(_MonoConfigureParams) AS=$(_MonoTuple)-as</_MonoConfigureParams>
+        <_MonoConfigureParams>$(_MonoConfigureParams) AR=$(_MonoTuple)-ar</_MonoConfigureParams>
+        <_MonoConfigureParams>$(_MonoConfigureParams) STRIP=$(_MonoTuple)-strip</_MonoConfigureParams>
```
Comment (lewurm) [inline src/mono/mono.proj]: where is `CC` and `CXX` set?

Comment (directhex) [reply src/mono/mono.proj]: https://github.com/dotnet/runtime/commit/f26a1cab799e

File: src/mono/mono.proj (hunk 1 above)
Comment (directhex) [reply src/mono/mono.proj]: Clang uses a `--target` flag for cross-compiling (rather than GCC where you have a `tuple-gcc` binary to emit for that target)

**Review by lewurm — APPROVED** [review]

File: src/mono/mono/mini/Makefile.am.in [hunk 3]
```diff
@@ -772,13 +772,13 @@ libmonoboehm_2_0_la_SOURCES =
...
 libmonoboehm_2_0_la_LIBADD = libmini.la $(interp_libs_with_mini) $(dbg_libs_with_mini) $(boehm_libs) $(LIBMONO_DTRACE_OBJECT) $(LLVMMONOF)
-libmonoboehm_2_0_la_LDFLAGS = $(libmonoldflags) $(monobin_platform_ldflags) 
+libmonoboehm_2_0_la_LDFLAGS = $(libmonoldflags) $(monobin_platform_ldflags) $(CCLDFLAGS)
 
 libmonosgen_2_0_la_SOURCES =
 libmonosgen_2_0_la_CFLAGS = $(mono_sgen_CFLAGS) @CXX_ADD_CFLAGS@
 
 libmonosgen_2_0_la_LIBADD = libmini.la $(interp_libs_with_mini) $(dbg_libs_with_mini) $(sgen_libs) $(LIBMONO_DTRACE_OBJECT) $(LLVMMONOF)
-libmonosgen_2_0_la_LDFLAGS = $(libmonoldflags) $(monobin_platform_ldflags) 
+libmonosgen_2_0_la_LDFLAGS = $(libmonoldflags) $(monobin_platform_ldflags) $(CCLDFLAGS)
```
Comment (akoeplinger) [inline src/mono/mono/mini/Makefile.am.in]: are we sure these are the only places that need this new `CCLDFLAGS` ?

Comment (directhex) [reply src/mono/mono/mini/Makefile.am.in]: For cross-compiling with clang, for the netcore subset: yes.

For all other potential subsets: I don't really care 🤷

File: src/mono/mono/profiler/Makefile.am [hunk 4]
```diff
@@ -8,6 +8,8 @@ AM_CPPFLAGS = \
 
 glib_libs = $(top_builddir)/mono/eglib/libeglib.la
 
+prof_ldflags = $(CCLDFLAGS)
```
Comment (lambdageek) [inline src/mono/mono/profiler/Makefile.am]: I think you need to change `prof_ldflags = ...` to `prof_ldflags += ...` lower down in this file in a couple of places

File: src/mono/mono.proj (hunk 1 above)
Comment (jaykrell) [reply src/mono/mono.proj]: Would be nice to somehow use either.

File: src/mono/mono.proj [hunk 5]
```diff
@@ -32,9 +32,29 @@
         <_MonoExtraCFLAGS>-O2 -g</_MonoExtraCFLAGS>
         <_MonoExtraCXXFLAGS>-O2 -g</_MonoExtraCXXFLAGS>
     </PropertyGroup>
+    <PropertyGroup Condition=" '$(TargetArchitecture)' == 'arm64' ">
+        <_MonoTuple>aarch64-linux-gnu</_MonoTuple>
+    </PropertyGroup>
+    <PropertyGroup Condition=" '$(TargetArchitecture)' == 'arm' ">
+        <_MonoTuple>arm-linux-gnueabihf</_MonoTuple>
+    </PropertyGroup>
+    <PropertyGroup Condition=" '$(TargetArchitecture)' == 'arm' Or '$(TargetArchitecture)' == 'arm64' ">
```
Comment (akoeplinger) [inline src/mono/mono.proj]: we should probably not set all these parameters when we're not cross compiling, e.g. if you directly build on an ARM machine

**Pull request merged**